java/dagger/internal/codegen/BindingGraph.java

```diff
@@ -1061,6 +1062,24 @@ boolean dependsOnLocalBindings(BindingKey bindingKey) {
         }
 
         /**
+         * Returns {@code true} if any of the bindings resolved for {@code bindingKey} are
+         * defined locally within this component's modules. If {@code bindingKey} is of kind
+         * {@link BindingKey.Kind#MEMBERS_INJECTION} each of the binding's dependencies will be
+         * checked to see if they are defined locally within this component's modules.
+         */
+        boolean bindingDefinedLocally(BindingKey bindingKey) {
+          if (bindingKey.kind().equals(BindingKey.Kind.MEMBERS_INJECTION)) {
+            return !getPreviouslyResolvedBindings(bindingKey)
+                .get()
+                .bindings()
+                .stream()
+                .flatMap(binding -> binding.dependencies().stream())
+                .anyMatch(dependency -> getLocalExplicitBindings(dependency.key()).isEmpty());
+          }
+          return !getLocalExplicitBindings(bindingKey.key()).isEmpty();
```

Comment:
Getting closer with ab5dbf1e.  This still isn't quiet right since bindings in the child are resolving bindings that should be unresolvable in the parent. Specifically `GraphValidationTest.bindingUsedOnlyInSubcomponentDependsOnBindingOnlyInSubcomponent()`